ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py

```diff
@@ -0,0 +1,230 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import copy
+from ansible_collections.arista.avd.plugins.filter.convert_dicts import convert_dicts
+from ansible_collections.arista.avd.plugins.module_utils.utils import get_all
+from ansible_collections.arista.avd.plugins.plugin_utils.schema.avdvalidator import AVD_META_SCHEMA
+
+
+try:
+    import jsonschema.validators
+    import jsonschema._validators
+    import jsonschema._types
+    import jsonschema
+except ImportError as imp_exc:
+    JSONSCHEMA_IMPORT_ERROR = imp_exc
+else:
+    JSONSCHEMA_IMPORT_ERROR = None
+
+try:
+    from deepmerge import always_merger
+except ImportError as imp_exc:
+    DEEPMERGE_IMPORT_ERROR = imp_exc
+else:
+    DEEPMERGE_IMPORT_ERROR = None
+
+
+def _keys(validator, keys, instance, schema):
+    '''
+    This function performs conversion on each key with the relevant subschema
+    '''
+    if not validator.is_type(instance, "object"):
+        return
+
+    # Don't run validator if $ref is part of the schema.
+    # Instead $ref validator will pop $ref and run all validators.
+    if "$ref" in schema:
+        return
+
+    # Resolve $ref for child keys, to support schema actions below which operates on the child schema
+    for key, childschema in keys.items():
+        if key in instance and "$ref" in childschema:
+            scope, resolved = validator.resolver.resolve(childschema["$ref"])
+            merged_childschema = copy.deepcopy(resolved)
+            always_merger.merge(merged_childschema, childschema)
+            merged_childschema.pop("$ref", None)
+            keys[key] = merged_childschema
+
+    # Perform type conversion of instance children if required based on "convert_types"
+    for key, childschema in keys.items():
+        if key in instance and "convert_types" in childschema:
+            instance[key] = _convert_types(validator, childschema["convert_types"], instance[key], childschema)
+
+    # Perform regular validation of each child element.
+    for key, childschema in keys.items():
+        if key in instance:
+            yield from validator.descend(
+                instance[key],
+                childschema,
+                path=key,
+                schema_path=key,
+            )
+
+    # Compile "dynamic_keys"
+    dynamic_keys = {}
+    schema_dynamic_keys = schema.get('dynamic_keys', {})
+    for dynamic_key, childschema in schema_dynamic_keys.items():
+        key_names = get_all(instance, dynamic_key)
+        for key_name in key_names:
+            if key_name not in dynamic_keys:
+                dynamic_keys[key_name] = childschema
```

Comment:
```suggestion
        for key_name in set(key_names) - set(dynamic_keys):
              dynamic_keys[key_name] = childschema
```

or (setdefault will only set a value if it does not exist)

```suggestion
        for key_name in key_names:
              dynamic_keys.setdefault(key_name, childschema)
```

Reply:
setdefault works. The other one doesn't. since we update the dynamic_keys inside the loop.

Reply:
fixed